rfc/20250728-opentofu-providers.md

```diff
@@ -0,0 +1,157 @@
+
+This RFC represents **one cohesive proposal** for how these components could work together, but I explicitly designed it to be flexible and evolutionary. The OpenTofu ecosystem benefits from multiple perspectives and approaches.
+
+In order to make this RFC easier to read and implement, I have split it into several focused documents:
+
+### Core Architecture
+
+1. [Provider Protocol](./20250728-opentofu-providers/01-provider-protocol.md) - Defines the stdio-based communication protocol using MessagePack
+2. [Provider Client Library](./20250728-opentofu-providers/02-provider-client-library.md) - Multiplexing library that abstracts protocol differences
+
+### Developer Experience
+
+3. [Provider SDK](./20250728-opentofu-providers/03-provider-sdk.md) - Language-specific SDKs with simple, idiomatic APIs
+4. [Local Execution](./20250728-opentofu-providers/04-local-execution.md) - Configuration and execution model for cmd+args providers
+
+### Distribution and Discovery
+
+5. [Registry Integration](./20250728-opentofu-providers/05-registry-integration.md) - Registry distribution and discovery mechanisms
+
+### Advanced Features
+
+6. [Provider Extensions](./20250728-opentofu-providers/06-provider-extensions.md) - Extensibility framework and proposed extensions:
+   - [Middleware Integration](./20250728-opentofu-providers/06a-middleware.md) - Provider-served middleware for governance and compliance
+   - [State Management Enhancements](./20250728-opentofu-providers/06b-state-management.md) - Advanced state handling capabilities
+
+## Background
+
+Current provider development using terraform-plugin-framework presents significant barriers:
+
+- **Language lock-in**: Providers must be written in Go, excluding developers from other ecosystems
+- **Complex toolchain**: Requires GoReleaser, GPG signing, GitHub Actions workflows
+- **Slow development cycles**: 15+ minute test feedback loops due to integration test requirements
+- **Excessive ceremony**: Scaffolding, boilerplate, and complex local testing setup
+
+These barriers prevent many engineers from creating custom providers for governance, internal tooling, or specialized use cases.
+
+## Proposed Solution
+
+OpenTofu Providers introduce:
+
+1. **Multi-language support**: SDKs for TypeScript, Python, Go, and other languages
+2. **Simple execution model**: Providers run as local processes and talk over simple protocols and transports
+3. **Progressive distribution**: Simplify the process from Local development → package managers → registry publication
+4. **Unified interface**: Seamless integration with existing provider ecosystem
+
+## Security Considerations
+
+OpenTofu providers run with the same security model as existing Terraform providers - they execute with the permissions of the user running the `tofu` command and have access to the same system resources.
```

Comment:
Given some of your other examples show running providers in containers, is that something that could be mentioned as an additional security measure?